src/dragonball/src/resource_manager.rs

```diff
@@ -0,0 +1,683 @@
+#[derive(Debug, PartialEq, thiserror::Error)]
+pub enum ResourceError {
+    /// Unknown/unsupported resource type.
+    #[error("unsupported resource type")]
+    UnknownResourceType,
+
+    /// Invalid resource range.
+    #[error("invalid resource range")]
+    InvalidResourceRange,
+
+    /// No source available.
+    #[error("no resource available")]
+    NoAvailResource,
+}
+
+struct ResoureManagerBuilder {
+    legacy_irq_pool: IntervalTree<()>,
+    msi_irq_pool: IntervalTree<()>,
+    pio_pool: IntervalTree<()>,
+    mmio_pool: IntervalTree<()>,
+    mem_pool: IntervalTree<()>,
+    kvm_mem_slot_pool: IntervalTree<()>,
+}
+
+impl Default for ResoureManagerBuilder {
+    fn default() -> Self {
+        ResoureManagerBuilder {
+            legacy_irq_pool: IntervalTree::new(),
+            msi_irq_pool: IntervalTree::new(),
+            pio_pool: IntervalTree::new(),
+            mmio_pool: IntervalTree::new(),
+            mem_pool: IntervalTree::new(),
+            kvm_mem_slot_pool: IntervalTree::new(),
+        }
+    }
+}
+
+impl ResoureManagerBuilder {
+    /// init legacy_irq_pool with arch specific constants.
+    fn init_legacy_irq_pool(mut self) -> Self {
+        // The LEGACY_IRQ_BASE irq is reserved for shared IRQ and won't be allocated / reallocated,
+        // so we don't insert it into the legacy_irq interval tree.
+        self.legacy_irq_pool
+            .insert(Range::new(LEGACY_IRQ_BASE + 1, LEGACY_IRQ_MAX), None);
+        self
+    }
+
+    /// init msi_irq_pool with arch specific constants.
+    fn init_msi_irq_pool(mut self) -> Self {
+        self.msi_irq_pool
+            .insert(Range::new(MSI_IRQ_BASE, MSI_IRQ_MAX), None);
+        self
+    }
+
+    /// init pio_pool with arch specific constants.
+    fn init_pio_pool(mut self) -> Self {
+        self.pio_pool.insert(Range::new(PIO_MIN, PIO_MAX), None);
+        self
+    }
+
+    /// Create mmio_pool with arch specific constants.
+    /// allow(clippy) is because `GUEST_MEM_START > MMIO_LOW_END`, we may modify GUEST_MEM_START or
+    /// MMIO_LOW_END in the future.
+    #[allow(clippy::absurd_extreme_comparisons)]
+    fn init_mmio_pool_helper(mmio: &mut IntervalTree<()>) {
+        mmio.insert(Range::new(MMIO_LOW_START, MMIO_LOW_END), None);
+        if !(*GUEST_MEM_END < MMIO_LOW_START
+            || GUEST_MEM_START > MMIO_LOW_END
+            || MMIO_LOW_START == MMIO_LOW_END)
+        {
+            #[cfg(target_arch = "x86_64")]
+            {
+                // Reserve the 64MB MMIO address range just below 4G, x86 systems have special
+                // devices, such as LAPIC, IOAPIC, HPET etc, in this range. And we don't explicitly
+                // allocate MMIO address for those devices.
+                // Reserve 16KB more for passthrough ipi.
+                let constraint = Constraint::new(0x400_4000u64)
+                    .min(0x1_0000_0000u64 - 0x400_4000u64)
+                    .max(0xffff_ffffu64);
```

Comment:
Could you define some `const`'s to help explain these values and avoid repetition?

Reply:
Added.